File: solutions/0229_majority-element-ii.java
// Problem: https://leetcode.com/problems/majority-element-ii
// Runtime: 19 ms

import java.util.ArrayList;
import java.util.HashSet;


class Solution {
    public List<Integer> majorityElement(int[] nums) {
        return new ArrayList<Integer>(majorityHelper(nums, 0, nums.length));
    }

    // returns, if it exists, all 1/3 majority element
    // start: inclusive, end: exclusive
    public Set<Integer> majorityHelper(int[] nums, int start, int end) {
        Set<Integer> out = new HashSet<>();
        int len = end - start;
        if (len < 3) { // base case: for len 0, 1, 2, include all
            for (int i = start; i < end; i++) {
                out.add(nums[i]);
            }
            return out;
        }
        // otherwise, split into 3 and recurse
        // 1 -> 1, 2 -> 1, 3 -> 1, 4 -> 2 -- ceil(n / 3)
        int partition = (len + 2) / 3;
        Set<Integer> l1 = majorityHelper(nums, start, start + partition);
        Set<Integer> l2 = majorityHelper(nums, start + partition, start + 2 * partition);
        Set<Integer> l3 = majorityHelper(nums, start + 2 * partition,
                                            Math.min(end, start + 3 * partition));
        for (Integer i : checkMajority(l1, nums, start, end)) {
            out.add(i);
        }
        for (Integer i : checkMajority(l2, nums, start, end)) {
            out.add(i);
        }
        for (Integer i : checkMajority(l3, nums, start, end)) {
            out.add(i);
        }
        return out;
    }

    public Set<Integer> checkMajority(Set<Integer> options, int[] nums, int start, int end) {
        int rec = (end - start) / 3; // minimum to be considered majority
        Set<Integer> out = new HashSet<>();
        for (Integer o : options) {
            int count = 0;
            for (int i = start; i < end; i++) {
                if (nums[i] == o) {
                    count++;
                }
            }
            if (count > rec) {
                out.add(o);
            }
        }
        return out;
    }
}

File: solutions/0229_majority-element-ii.py
# Problem: https://leetcode.com/problems/majority-element-ii
# Runtime: 113 ms

class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        if len(nums) < 2:
            return nums
        if len(nums) == 2:
            return nums if nums[0] != nums[1] else [nums[0]]

        hashMap = {}
        for num in nums:
            hashMap[num] = hashMap.get(num, 0) + 1

        vals = [x for x in hashMap.items()]
        if len(vals) == 1:
            return [vals[0][0]]
        max1 = int(vals[0][1] < vals[1][1])
        max2 = 1 - max1
        for i in range(2, len(vals)):
            curr = vals[i][1]
            if curr > vals[max1][1]:
                max2 = max1
                max1 = i
            elif curr > vals[max2][1]:
                max2 = i
        
        requiredSize = len(nums) // 3
        print(vals)
        print(requiredSize)
        if vals[max2][1] > requiredSize:
            return [vals[max1][0], vals[max2][0]]
        if vals[max1][1] > requiredSize:
            return [vals[max1][0]]
        return []